### PR TITLE
#545 fixing SyncIterable thread-safety bug

### PR DESCRIPTION
**Pull request by svendiedrichsen**
Replaced SyncIterable extending IterableEnvelope with implementing Iterable<> interface and synchronized access to iterator method

### CODE REVIEW

**Comment by 0crat** [comment]
@yegor256[/z](https://www.0crat.com/u/yegor256) please, pay attention to this pull request

**Comment by 0crat** [comment]
Job [#577](https://github.com/yegor256/cactoos/issues/577) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#577](https://github.com/yegor256/cactoos/issues/577) is assigned to @fabriciofx[/z](https://www.0crat.com/u/fabriciofx). The budget is 15 minutes, see [§4](http://datum.zerocracy.com/pages/policy.html#4). Please, read [§27](http://datum.zerocracy.com/pages/policy.html#27) and when you decide to accept the changes, inform @yegor256[/z](https://www.0crat.com/u/yegor256) (the architect) right in this ticket. If you decide that this PR should not be accepted ever, also inform the architect.

**Comment by svendiedrichsen** [comment]
@fabriciofx Please review.

File: src/main/java/org/cactoos/iterable/SyncIterable.java [hunk 1]
```diff
@@ -58,7 +64,16 @@ public SyncIterable(final X... src) {
      * @param iterable The iterable
      */
     public SyncIterable(final Iterable<X> iterable) {
-        super(new SyncScalar<>(() -> iterable));
+        this.iterable = new UncheckedScalar<>(
```
Comment (fabriciofx) [inline src/main/java/org/cactoos/iterable/SyncIterable.java]: @svendiedrichsen Please, don't use the same attribute and parameter name. I recommend you change `this.iterable` to `this.origin`.

File: src/main/java/org/cactoos/iterable/SyncIterable.java [hunk 2]
```diff
@@ -58,7 +64,16 @@ public SyncIterable(final X... src) {
...
+            () -> iterable
+        );
+    }
+
+    @Override
+    public Iterator<X> iterator() {
+        synchronized (this.iterable) {
```
Comment (fabriciofx) [inline src/main/java/org/cactoos/iterable/SyncIterable.java]: @svendiedrichsen Why are you using the same object as a lock? Please, follow the same approach used by Yegor in `SyncInput` (a attribute to lock, and a second ctor).

**Comment by fabriciofx** [comment]
@yegor256 This PR seems fine to me. I accept it.

**Comment by yegor256** [comment]
@rultor merge

**Comment by 0crat** [comment]
Order was successfully finished: +15 points just awarded to @fabriciofx[/z](https://www.0crat.com/u/fabriciofx), total is [+390](https://www.0crat.com/u/fabriciofx)

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://datum.zerocracy.com/pages/policy.html#28): +15 points just awarded to @yegor256[/z](https://www.0crat.com/u/yegor256), total is [+8550](https://www.0crat.com/u/yegor256)

**Comment by 0crat** [comment]
The job [#577](https://github.com/yegor256/cactoos/issues/577) is now out of scope

**Pull request merged**